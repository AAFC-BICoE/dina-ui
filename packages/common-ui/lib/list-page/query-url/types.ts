export interface SimpleQueryGroup {
  // Conjunction (AND/OR)
  c: string;

  // Props: Items inside the group
  p: SimpleQueryRow[];
}

export interface SimpleQueryRow {
  // Field
  f: string;

  // Operator
  o: string;

  // Value
  v: string;

  // Type
  t: string;

  // Dynamic field identifier (UUID for example if managed attribute)
  d?: string;
}
