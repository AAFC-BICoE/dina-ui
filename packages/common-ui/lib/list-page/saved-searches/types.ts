import { JsonTree } from "react-awesome-query-builder";

/**
 * Since saved searches are stored as JSON this is the structure to be used.
 */
export interface SavedSearchStructure {
  [indexName: string]: {
    [savedSearchName: string]: SingleSavedSearch;
  };
}

export interface SingleSavedSearch {
  // This field is not saved in the JSON.
  savedSearchName?: string;
  default: boolean;
  queryTree?: JsonTree;
}
