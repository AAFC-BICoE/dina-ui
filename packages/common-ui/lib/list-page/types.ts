import { KitsuResource } from "kitsu";
import { Column } from "react-table";

/**
 * This type extends the react-table column type, this just adds a few specific fields for elastic
 * search mapping and internationalization.
 */
export interface TableColumn<TData extends KitsuResource>
  extends Column<TData> {
  /**
   * User-friendly column to be displayed. You can use a DinaMessage key for internationalization.
   */
  label?: string;

  /**
   * Elastic search path to the attribute.
   *
   * Example: `data.attributes.name`
   */
  attributePath?: string;

  /**
   * This field is used to find the relationship in the included section.
   */
  relationshipType?: string;

  /**
   * Is this attribute considered a keyword in elastic search. Required for filtering and sorting.
   */
  isKeyword?: boolean;

  /**
   * The QueryPage will only display the accessors that are displayed on the result table. However,
   * if you have custom cells that receive other fields you will need to add them to this list so
   * elastic search includes the fields in the result.
   *
   * Example: `data.attributes.name`
   *
   * Please note that duplicate fields are automatically removed so you don't need to worry about
   * having unique accessors.
   */
  additionalAccessors?: string[];
}

/**
 * The full path will be generated for elastic using a combination of the parent path and
 * the value. The path is generated using the following:
 *
 * {parentPath}.{path}.{value}
 *
 * Example: included.attributes.determination.verbatimScientificName
 */
export interface ESIndexMapping {
  /**
   * Name of the attribute.
   *
   * Example: verbatimScientificName
   */
  value: string;

  /**
   * Text that is displayed to the user in the Query Filtering option menu.
   *
   * This text is a user-friendly generated label, which may show some paths to help the user
   * understand the relationships better. This is generated from the path.
   *
   * Example: determination.verbatimScientificName
   */
  label: string;

  /**
   * The attributes type. This can change how the query row is displayed and the options provided.
   *
   * Examples: text, keyword, boolean, date, boolean, long, short, integer...
   */
  type: string;

  /**
   * If enabled, it will allow the user to see suggestions as they type. The suggestions will come
   * from elastic search based on most common values saved.
   *
   * Only available for the text type.
   */
  distinctTerm: boolean;

  /**
   * If enabled it will allow the user to search based on the starting of a word.
   *
   * Example: Hexapoda can be matched with "hex".
   */
  startsWithSupport: boolean;

  /**
   * If enabled it will allow the user to search based in the middle of a word.
   *
   * Example: Hexapoda can be matched with "pod".
   */
  containsSupport: boolean;

  /**
   * If enabled it will allow the user to search based on the ending of a word.
   *
   * Example: Hexapoda can be matched with "poda".
   */
  endsWithSupport: boolean;

  /**
   * The path for the attribute without the attribute name. This path does not include the parent
   * path.
   *
   * Example: attribute.determination
   */
  path: string;

  /**
   * If the attribute belongs to a relationship, this is the path for only the parent. When generating
   * the elastic search query it will use this as the prefix of the path.
   *
   * Example: included
   */
  parentPath?: string;

  /**
   * If the attribute belongs to a relationship, this is the name which will be used to group
   * attributes under the same relationship together in the search. This name will also be used to
   * display text of the group.
   *
   * This text should match the user-friendly label in the locales files. It will be searched using
   * title_[parentName].
   *
   * Example: preparationMethod (and the label will be `title_preparationMethod`)
   */
  parentName?: string;

  /**
   * The parent type is the relationship type to be used. This will be used for elastic search
   * filtering. The reason this is used since the same field can be used for multiple relationships.
   *
   * Example: preparation-method
   */
  parentType?: string;

  /**
   * Only provided if it was added using a dynamic field config.
   */
  dynamicField?: DynamicField;
}

/**
 * Each type on the query builder has a function that is used to transform the query builder row
 * into elastic search logic.
 *
 * @see QueryBuilderElasticSearchExport
 * @see QueryRowTextSearch
 */
export interface TransformToDSLProps {
  /**
   * The query builder type, not really used but required in the function for the QueryBuilder
   * library.
   *
   * For the Elastic Search type you can use the fieldInfo.type instead.
   */
  queryType: string;

  /**
   * The value to search against. If the type is "text" then this is the text that will be filtered.
   */
  value: string;

  /**
   * The operation being performed.
   *
   * For example:
   * "equals", "notEquals", "containsText"
   *
   * Operators are defined in the QueryBuilderConfig file.
   */
  operation: string;

  /**
   * This field path is the unique path given to each item in the field list.
   *
   * For getting the proper absolute field path, use the FieldInfo to generate it.
   */
  fieldPath: string;

  /**
   * The elastic search mapping for the field.
   */
  fieldInfo?: ESIndexMapping;
}

export type DynamicFieldType = "managedAttribute" | "fieldExtension";

export interface DynamicFieldsMappingConfig {
  /** Attribute level dynamic fields */
  fields: DynamicField[];

  /** Dynamic fields for relationships */
  relationshipFields: RelationshipDynamicField[];
}

export interface DynamicField {
  /**
   * Option label that should be used.
   */
  label: string;

  type: DynamicFieldType;

  path: string;

  /**
   * Endpoint where these dynamic fields can be retrieved to list.
   *
   * Example: "collection-api/managed-attribute"
   */
  apiEndpoint: string;

  /**
   * Optional field to indicate which Managed Attributes or Field Extensions should be listed.
   */
  component?: string;
}

/**
 * Configuration for where the Dynamic Field can be found within the relationship index mapping.
 */
export interface RelationshipDynamicField extends DynamicField {
  referencedBy: string;
  referencedType: string;
}
