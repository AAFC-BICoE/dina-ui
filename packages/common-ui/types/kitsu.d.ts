/// <reference types="node" />

import { AxiosInstance } from "axios";
import { SetRequired } from "type-fest";

declare module "kitsu" {
  // export default Kitsu;

  /** JSONAPI client. */
  export default class Kitsu {
    public axios: AxiosInstance;

    public headers: any;

    constructor(params: KitsuConstructorParams);

    public delete(...args: any[]): Promise<any>;

    public get<TData, TMeta = undefined>(
      path: string,
      params: GetParams
    ): Promise<KitsuResponse<TData, TMeta>>;

    public patch(...args: any[]): Promise<any>;

    public post(...args: any[]): Promise<any>;
  }

  /** Params for the Kitsu JSONAPI client's constructor */
  export interface KitsuConstructorParams {
    baseURL: string;
    headers?: any;
    pluralize?: boolean;
    camelCaseTypes?: boolean;
    resourceCase?: "kebab" | "snake" | "none";
  }

  /** Parameters for GET requests. */
  export interface GetParams {
    /** Fields to include in the response data. */
    fields?: FieldsParam;

    /** Resource filter */
    filter?: FilterParam;

    /**
     * Sort order + attribute.
     * Examples:
     *  - name
     *  - -description
     */
    sort?: string;

    /** Included resources. */
    include?: string;

    /** Vendor-specific parameter for paginating listed data. */
    page?: any;
  }

  /** Parameter for requesting sparse fields. */
  export interface FieldsParam {
    [key: string]: string;
  }

  /** Parameter for filtering listed data. */
  export type FilterParam = string | Record<string, string>;

  /** The response from a Kitsu GET request. */
  export interface KitsuResponse<
    TData extends KitsuResponseData,
    TMeta = undefined
  > {
    data: TData extends (infer R)[]
      ? PersistedResource<R>[]
      : PersistedResource<TData>;
    meta: TMeta;
  }

  /** The Kitsu response data can either be one resource or an array of resources. */
  export type KitsuResponseData = KitsuResource | KitsuResource[];

  /** JSONAPI resource base attributes. */
  export interface KitsuResource {
    id?: string;
    type: string;
  }

  export interface KitsuResourceLink {
    id: string | null;
    type: string;
  }

  /**
   * Makes the 'id' field required on a resource type and all of its relationships.
   * Used when assuming that data from the back-end always has the ID set.
   */
  export type PersistedResource<TData extends KitsuResource = KitsuResource> = {
    [P in keyof TData]: TData[P] extends KitsuResource
      ? PersistedResource<TData[P]>
      : TData[P] extends KitsuResource | undefined
      ? PersistedResource<TData[P]> | undefined
      : TData[P];
  } &
    Required<KitsuResource>;

  /**
   * Used when creating or updating a resource.
   * Makes the 'id' field optional on the main resource.
   * Makes the 'id' field required on linked resources.
   * 'type' must be defined.
   */
  export type InputResource<TData extends KitsuResource> = SetRequired<
    {
      [P in keyof TData]?: TData[P] extends KitsuResource | undefined
        ? Required<KitsuResourceLink> | undefined
        : TData[P];
    },
    "type"
  >;
}
