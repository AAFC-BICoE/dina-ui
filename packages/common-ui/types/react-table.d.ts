import "@tanstack/react-table";

declare module "@tanstack/table-core" {
  interface ColumnMeta<TData extends RowData, TValue> {
    className?: string;
    style?: CSSProperties;
  }

  interface TableMeta<TData extends RowData> {
    updateData?: (rowIndex: number, columnId: string, value: unknown) => void;
    enableEditing?: boolean;
  }
}
