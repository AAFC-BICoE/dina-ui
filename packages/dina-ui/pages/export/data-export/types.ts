export interface SavedExportColumnStructure {
  name: string;
  component: string;
  columns: string[];
}
