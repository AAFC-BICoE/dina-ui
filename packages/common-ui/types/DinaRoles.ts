export const DINA_ADMIN = "dina-admin";
export const READ_ONLY_ADMIN = "read-only-admin";
export const SUPER_USER = "super-user";
export const USER = "user";
export const GUEST = "guest";
export const READ_ONLY = "read-only";
