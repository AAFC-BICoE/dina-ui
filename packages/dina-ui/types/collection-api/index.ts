export * from "./resources/Assemblage";
export * from "./resources/CollectingEvent";
export * from "./resources/Collection";
export * from "./resources/CollectorGroup";
export * from "./resources/FormTemplate";
export * from "./resources/Determination";
export * from "./resources/GeographicPlaceNameSourceDetail";
export * from "./resources/GeoReferenceAssertion";
export * from "./resources/Institution";
export * from "./resources/MaterialSample";
export * from "./resources/form-legends/MaterialSampleForm";
export * from "./resources/MaterialSampleRunConfig";
export * from "./resources/MaterialSampleType";
export * from "./resources/Organism";
export * from "./resources/Project";
export * from "./resources/Protocol";
export * from "./resources/PreparationType";
export * from "./resources/PreparationMethod";
export * from "./resources/StorageUnit";
export * from "./resources/StorageUnitType";
export * from "./resources/VocabularyElement";
export * from "./resources/CollectionMethod";
export * from "./resources/ProtocolElement";
export * from "./resources/ManagedAttribute";
