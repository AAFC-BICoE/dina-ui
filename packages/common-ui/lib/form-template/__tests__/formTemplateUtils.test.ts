import { FormTemplate } from "../../../../dina-ui/types/collection-api";
import {
  getFormTemplateComponent,
  getFormTemplateField,
  getFormTemplateSection
} from "../formTemplateUtils";

const COMPONENT_NAME = "component1";
const SECTION_NAME = "section1";
const FIELD_NAME = "field1";

const FORM_TEMPLATE_TEST: FormTemplate = {
  type: "form-template",
  components: [
    {
      name: COMPONENT_NAME,
      visible: true,
      order: 0,
      sections: [
        {
          name: SECTION_NAME,
          gridPositionX: 0,
          gridPositionY: 0,
          visible: true,
          items: [
            {
              name: FIELD_NAME,
              gridPositionX: 0,
              gridPositionY: 0,
              defaultValue: "defaultValue1",
              visible: true
            }
          ]
        }
      ]
    },
    {
      name: "anotherComponentName",
      visible: true,
      order: 0,
      sections: [
        {
          name: "anotherComponentSection",
          gridPositionX: 0,
          gridPositionY: 0,
          visible: true,
          items: [
            {
              name: "anotherComponentField",
              gridPositionX: 0,
              gridPositionY: 0,
              defaultValue: undefined,
              visible: true
            }
          ]
        }
      ]
    }
  ]
};

describe("Form Template Utils", () => {
  describe("getFormTemplateComponent helper function", () => {
    test("If any of the props are undefined, it should return undefined", () => {
      expect(
        getFormTemplateComponent(undefined, COMPONENT_NAME)
      ).toBeUndefined();

      expect(
        getFormTemplateComponent(FORM_TEMPLATE_TEST, undefined)
      ).toBeUndefined();
    });

    test("If the component doesn't exist in the form template, it should return undefined", () => {
      expect(
        getFormTemplateComponent(FORM_TEMPLATE_TEST, "incorrectComponentName")
      ).toBeUndefined();
    });

    test("If the component does exist, the component properties should be returned", () => {
      const componentProps = getFormTemplateComponent(
        FORM_TEMPLATE_TEST,
        COMPONENT_NAME
      );

      // It should exist.
      expect(componentProps).toBeTruthy();

      // Check to ensure the correct field was returned.
      expect(componentProps).toMatchSnapshot();
    });
  });

  describe("getFormTemplateSection helper function", () => {
    test("If any of the props are undefined, it should return undefined", () => {
      expect(
        getFormTemplateSection(undefined, COMPONENT_NAME, SECTION_NAME)
      ).toBeUndefined();

      expect(
        getFormTemplateSection(FORM_TEMPLATE_TEST, undefined, SECTION_NAME)
      ).toBeUndefined();

      expect(
        getFormTemplateSection(FORM_TEMPLATE_TEST, COMPONENT_NAME, undefined)
      ).toBeUndefined();
    });

    test("If the section doesn't exist in the form template, it should return undefined", () => {
      expect(
        getFormTemplateSection(
          FORM_TEMPLATE_TEST,
          "incorrectComponentName",
          SECTION_NAME
        )
      ).toBeUndefined();

      expect(
        getFormTemplateSection(
          FORM_TEMPLATE_TEST,
          COMPONENT_NAME,
          "incorrectSectionName"
        )
      ).toBeUndefined();
    });

    test("If the section does exist, the section properties should be returned", () => {
      const sectionProps = getFormTemplateSection(
        FORM_TEMPLATE_TEST,
        COMPONENT_NAME,
        SECTION_NAME
      );

      // It should exist.
      expect(sectionProps).toBeTruthy();

      // Check to ensure the correct field was returned.
      expect(sectionProps).toMatchSnapshot();
    });
  });

  describe("getFormTemplateField helper function", () => {
    test("If any of the props are undefined, it should return undefined", () => {
      expect(
        getFormTemplateField(
          undefined,
          COMPONENT_NAME,
          SECTION_NAME,
          FIELD_NAME
        )
      ).toBeUndefined();

      expect(
        getFormTemplateField(
          FORM_TEMPLATE_TEST,
          undefined,
          SECTION_NAME,
          FIELD_NAME
        )
      ).toBeUndefined();

      expect(
        getFormTemplateField(
          FORM_TEMPLATE_TEST,
          COMPONENT_NAME,
          undefined,
          FIELD_NAME
        )
      ).toBeUndefined();

      expect(
        getFormTemplateField(
          FORM_TEMPLATE_TEST,
          COMPONENT_NAME,
          SECTION_NAME,
          undefined
        )
      ).toBeUndefined();
    });

    test("If the field doesn't exist in the form template, it should return undefined", () => {
      expect(
        getFormTemplateField(
          FORM_TEMPLATE_TEST,
          "incorrectComponentName",
          SECTION_NAME,
          FIELD_NAME
        )
      ).toBeUndefined();

      expect(
        getFormTemplateField(
          FORM_TEMPLATE_TEST,
          COMPONENT_NAME,
          "incorrectSectionName",
          FIELD_NAME
        )
      ).toBeUndefined();

      expect(
        getFormTemplateField(
          FORM_TEMPLATE_TEST,
          COMPONENT_NAME,
          SECTION_NAME,
          "incorrectFieldName"
        )
      ).toBeUndefined();
    });

    test("If the field does exist, the field properties should be returned", () => {
      const fieldProps = getFormTemplateField(
        FORM_TEMPLATE_TEST,
        COMPONENT_NAME,
        SECTION_NAME,
        FIELD_NAME
      );

      // It should exist.
      expect(fieldProps).toBeTruthy();

      // Check to ensure the correct field was returned.
      expect(fieldProps).toMatchSnapshot();
    });
  });
});
