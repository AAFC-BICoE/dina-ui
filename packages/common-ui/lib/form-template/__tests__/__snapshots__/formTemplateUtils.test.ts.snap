// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Template Utils getFormTemplateComponent helper function If the component does exist, the component properties should be returned 1`] = `
Object {
  "name": "component1",
  "order": 0,
  "sections": Array [
    Object {
      "gridPositionX": 0,
      "gridPositionY": 0,
      "items": Array [
        Object {
          "defaultValue": "defaultValue1",
          "gridPositionX": 0,
          "gridPositionY": 0,
          "name": "field1",
          "visible": true,
        },
      ],
      "name": "section1",
      "visible": true,
    },
  ],
  "visible": true,
}
`;

exports[`Form Template Utils getFormTemplateField helper function If the field does exist, the field properties should be returned 1`] = `
Object {
  "defaultValue": "defaultValue1",
  "gridPositionX": 0,
  "gridPositionY": 0,
  "name": "field1",
  "visible": true,
}
`;

exports[`Form Template Utils getFormTemplateSection helper function If the section does exist, the section properties should be returned 1`] = `
Object {
  "gridPositionX": 0,
  "gridPositionY": 0,
  "items": Array [
    Object {
      "defaultValue": "defaultValue1",
      "gridPositionX": 0,
      "gridPositionY": 0,
      "name": "field1",
      "visible": true,
    },
  ],
  "name": "section1",
  "visible": true,
}
`;
